_drafts/cortex-m-fault-debug.md

```diff
@@ -0,0 +1,871 @@
+  // Clear any faults from the CFSR
+  *cfsr |= *cfsr;
+  // the instruction we will return to when we exit from the exception
+  frame->return_address = (uint32_t)recover_from_task_fault;
+  // the function we are returning too should never branch
+  // so set lr to a pattern that would fault if it did
+  frame->lr = 0xdeadbeef;
+  // reset the psr state and only leave the
+  // "thumb instruction interworking" bit set
+  frame->xpsr = (1 << 24);
+```
+
+You may recall from our [previous post](https://interrupt.memfault.com/blog/cortex-m-rtos-context-switching#context-state-stacking), fault handlers can work just like regular c functions so after these changes we will exit from `my_fault_handler_c` and start executing whatever is in `recover_from_task_fault`.
+
+> For those who want to walk through this actual example, check out the example with gdb [below](#usage-fault-recovery-example)
+
+## Faults from Faults!
+
+The astute observer might wonder what happens when a new fault occurs in the code dealing with a fault.
+If you have enabled configurable fault handlers (i.e MemoryManagement, BusFault, or UsageFault), a fault generated in these handlers will trigger a Hardfault.
+
+Once in the Hardfault Handler, the ARM Core is said to be operating at priority level -1. At this level or above, a fault will put the processor in an unrecoverable state where a reset is expected known as **Lockup**.
+
+Typically, the processor will automatically reset the processor upon entering lockup but this is not a requirement per the specification. For example, you may have to enable a hardware watchdog for a reset to take place. It's worth double checking the reference manual for the MCU being used for clarification.
+
+When a debugger is attached, lockup often has a different behavior. For example, on the NRF52840, "Reset from CPU lockup is disabled if the device is in debug interface mode".
+
+Technically when a lockup happens, the processor will repeatedly fetch the same fixed instruction, 0xFFFFFFFE or the address which triggered the fault, in a loop until a reset occurs.
+
+> Fun Fact: Whether or not some classes of MemManage Faults or BusFaults trigger a fault from an exception is actually configurable via the MPU_CTRL.HFNMIENA & CCR.BFHFNMIGN configurations.
+
+## Automating the Analysis
+
+### Instrumenting the code
+
+Many Real Time Operating Systems (**RTOS**) targetting Cortex-M devices will add options to dump verbose fault information to the console upon crash. Some examples include Arm Mbed OS[^2] and Zephyr[^3]. This approach has a couple notable limitations:
+
+- It bloats the code & data size of the binary image and consequently often gets disabled via compiler flags
+- It requires a firmware update to improve or fix issues with the analyzers
+- It requires an active console session be active to see what fault occurred and depending on the console setup in the system this can be flaky if the system is in a crashed state
```

Comment:
Maybe add just a log dump of a crash on one of those systems to see what kind of instrumentation we're talking about?